staging/src/k8s.io/apiserver/pkg/storage/etcd3/etcdfeature/feature_support_checker.go

```diff
@@ -0,0 +1,144 @@
+}
+
+type defaultFeatureSupportChecker struct {
+	// mutex protects progressNotifySupported and progresNotifyEndpointCache.
+	mux                        sync.Mutex
+	progressNotifySupported    bool
+	progresNotifyEndpointCache map[string]bool
+}
+
+func newDefaultFeatureSupportChecker() *defaultFeatureSupportChecker {
+	return &defaultFeatureSupportChecker{
+		progresNotifyEndpointCache: make(map[string]bool),
+		// When the progresNotifyEndpointCache is empty the should be true.
+		progressNotifySupported: true,
+	}
+}
+
+// Supports can check the featue from anywhere without storage if it was checked before.
+func (f *defaultFeatureSupportChecker) Supports(feature storage.Feature) (bool, error) {
+	switch feature {
+	case storage.RequestWatchProgress:
+		f.mux.Lock()
+		defer f.mux.Unlock()
+
+		return f.progressNotifySupported, nil
+	default:
+		return false, fmt.Errorf("feature %q is not implemented in DefaultFeatureSupportChecker", feature)
+	}
+}
+
+// ClientSupports accepts client and calcuate the support per endpoint and caches it.
+// It will return at any point if error happens or one endpoint is not supported.
+func (f *defaultFeatureSupportChecker) ClientSupports(ctx context.Context, c client, feature storage.Feature) (bool, error) {
+	f.mux.Lock()
+	defer f.mux.Unlock()
+
+	for _, ep := range c.Endpoints() {
+		supported, err := f.checkAndCacheEndpoint(ctx, c, ep)
+		if err != nil {
+			return false, err
+		}
+		if !supported {
+			f.progressNotifySupported = false
+			return false, nil
+		}
+	}
+	return true, nil
+}
+
+func (f *defaultFeatureSupportChecker) checkAndCacheEndpoint(ctx context.Context, c client, ep string) (bool, error) {
+	if supported, ok := f.progresNotifyEndpointCache[ep]; ok {
+		return supported, nil
+	}
+
+	supported, err := endpointSupportsRequestWatchProgress(ctx, c, ep)
+	if err != nil {
+		return false, err
+	}
+
+	f.progresNotifyEndpointCache[ep] = supported
+	return supported, nil
+}
+
+// client defines the interface required for testing purposes.
+type client interface {
+	// Endpoints returns list of endpoints in etcd client.
+	Endpoints() []string
+	// Status retrieves the status information from the etcd client connected to the specified endpoint.
+	// It takes a context.Context parameter for cancellation or timeout control.
+	// It returns a clientv3.StatusResponse containing the status information or an error if the operation fails.
+	Status(ctx context.Context, endpoint string) (*clientv3.StatusResponse, error)
+}
+
+// endpointSupportsRequestWatchProgress evaluates whether RequestWatchProgress supported by current version of etcd endpoint.
+// Based on this issues:
+//   - Issue #15220: https://github.com/etcd-io/etcd/issues/15220
+//   - KEP 2340: Consistent reads from cache: https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/2340-Consistent-reads-from-cache/README.md#bug-in-etcd-progress-notification
```

Comment:
```suggestion
//   - https://github.com/etcd-io/etcd/issues/15220 - Fixed in etcd v3.4.25+ amd v3.5.8+
//   - https://github.com/etcd-io/etcd/issues/17507 - Fixed in etcd v3.4.31+ and v3.5.13+
```

Mentioning KEP here is redundant.